### PR TITLE
Scenario deduction

### PR DESCRIPTION
**Pull request by jialuw96**
## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by jialuw96** [comment]
@adowling2 Hi Alex, the code is ready for review. In Fim_doe.py the compute_FIM function is changed, containing sequential_finite and direct_kaug modes. In reactor_kinetics_parmest.py is the most recent create_model function needed by pyomo.DOE. Thank you!

File: pyomo/contrib/doe/fim_doe.py [hunk 1]
```diff
@@ -205,26 +208,38 @@ def __generate_flatten_timeset(self, all_info, flatten_measure_name,name_and_ind
         self.flatten_measure_timeset = flatten_measure_timeset
 
     def __model_measure_name(self):
-        '''Return pyomo string name 
+        '''Return pyomo string name
```
Comment (adowling2) [inline pyomo/contrib/doe/fim_doe.py:211]: This function does not return anything. Instead, it updates self.measurement_names_ef and self.measurement_names_kaug.

File: pyomo/contrib/doe/fim_doe.py [hunk 2]
```diff
@@ -205,26 +208,38 @@ def __generate_flatten_timeset(self, all_info, flatten_measure_name,name_and_ind
...
             # check if there is extra index
             if self.ind_string in mname:
                 measure_name = mname.split(self.ind_string)[0]
                 measure_index = mname.split(self.ind_string)[1]
+
                 for tim in self.flatten_measure_timeset[mname]:
                     # get the measurement name in the model
-                    measurement_name = measure_name + '[0,' + measure_index + ',' + str(tim) + ']'
-                    measurement_names.append(measurement_name)
+                    measurement_name_kaug = measure_name + '[' + measure_index + ',' + str(tim) + ']'
```
Comment (adowling2) [inline pyomo/contrib/doe/fim_doe.py:228]: Add comments showing an example of a measurement name for k_aug and ef

File: pyomo/contrib/doe/fim_doe.py [hunk 3]
```diff
@@ -205,26 +208,38 @@ def __generate_flatten_timeset(self, all_info, flatten_measure_name,name_and_ind
...
+                    measurement_names_kaug.append(measurement_name_kaug)
+
+                    measurement_name_ef = measure_name + "['" + measure_index + "'," + str(tim) + "]"
+                    measurement_names_ef.append(measurement_name_ef)
+
+
             else:
```
Comment (adowling2) [inline pyomo/contrib/doe/fim_doe.py:235]: Add a comment "if there is no extra index"

File: pyomo/contrib/doe/fim_doe.py [hunk 4]
```diff
@@ -205,26 +208,38 @@ def __generate_flatten_timeset(self, all_info, flatten_measure_name,name_and_ind
...
                 for tim in self.flatten_measure_timeset[mname]:
                     # get the measurement name in the model
-                    measurement_name = mname + '[0,' + str(tim) + ']'
-                    measurement_names.append(measurement_name)
-        self.model_measure_name = measurement_names
+                    measurement_name = mname + '[' + str(tim) + ']'
+                    measurement_names_ef.append(measurement_name)
```
Comment (adowling2) [inline pyomo/contrib/doe/fim_doe.py:239]: Add a comment to explain why these are the same. I think I understand what you are doing. You want comments to make it explicit so there is no need for the reader to guess.

File: pyomo/contrib/doe/fim_doe.py [hunk 5]
```diff
@@ -268,9 +283,9 @@ def SP_measure_name(self, j, t,scenario_all=None, p=None, mode=None, legal_t=Tru
                 if type(self.name_and_index[measure_name][0]) is str:
```
Comment (adowling2) [inline pyomo/contrib/doe/fim_doe.py:283]: This function needs comments before each if and else line. I can probably figure it out with enough time. But a few good comments prevent me from needing to guess.

File: pyomo/contrib/doe/fim_doe.py [hunk 6]
```diff
@@ -471,12 +486,15 @@ def optimize_doe(self,  design_values, if_optimize=True, objective_option='det',
...
-            self.timepoint_overall_set = jac_involved_measurement.timepoint_overall_set.copy()
+            #self.jac_involved_name = jac_involved_measurement.flatten_measure_name.copy()
+            self.jac_involved_name = copy.deepcopy(jac_involved_measurement.flatten_measure_name)
+            #self.timepoint_overall_set = jac_involved_measurement.timepoint_overall_set.copy()
+            self.timepoint_overall_set = copy.deepcopy(jac_involved_measurement.timepoint_overall_set)
         else:
-            self.jac_involved_name = self.flatten_measure_name.copy()
-            self.timepoint_overall_set = self.measure.timepoint_overall_set.copy()
-            
+            #self.jac_involved_name = self.flatten_measure_name.copy()
```
Comment (adowling2) [inline pyomo/contrib/doe/fim_doe.py:494]: Add a comment such as "if jac_involved_measurement is None, assume..."

File: pyomo/contrib/doe/fim_doe.py [hunk 7]
```diff
@@ -623,6 +641,9 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
             # call generator function to get scenario dictionary
             scena_gen = Scenario_generator(self.param_init, formula=self.formula, step=self.step)
             scena_gen.generate_sequential_para()
+            scena_set = scena_gen.simultaneous_scenario()
+            self.scena_set = scena_set
+            print(self.scena_set)
```
Comment (adowling2) [inline pyomo/contrib/doe/fim_doe.py:646]: You should have an integer in self.print_level that toggles on print statements such as 0 = no extra print statements, 1 = a few extra print statements, 2 = many extra print statements. For any extra print statements, you need print out extra text to the screen, not just a variable.

File: pyomo/contrib/doe/fim_doe.py [hunk 8]
```diff
@@ -898,6 +912,7 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
             scena_gen = Scenario_generator(self.param_init, formula=None, step=self.step)
             scenario_all = scena_gen.simultaneous_scenario()
 
+            print(scenario_all)
```
Comment (adowling2) [inline pyomo/contrib/doe/fim_doe.py:915]: See comment above about print levels.

File: pyomo/contrib/doe/fim_doe.py [hunk 9]
```diff
@@ -934,16 +950,17 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
 
             # set ub and lb to parameters
             for par in self.param_name:
-                component = eval('mod.'+par+'[0]')
+                #component = eval('mod.'+par+'[0]')
+                component = eval('mod.'+par)
```
Comment (adowling2) [inline pyomo/contrib/doe/fim_doe.py:954]: Is there a safer way to do this in Pyomo? What does parmest do?

File: pyomo/contrib/doe/fim_doe.py [hunk 10]
```diff
@@ -970,11 +987,15 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
...
 
             # loop over measurement variables and their time points
-            for measurement_name in self.measure.model_measure_name:
+            for measurement_name in self.measure.model_measure_name_kaug:
                 # get right line number in kaug results
+                print('name:', measurement_name)
                 if self.discretize_model is not None:
                     # for DAE model, some variables are fixed
                     try:
+                    #if measurement_name == 'C[CA,0]':
```
Comment (adowling2) [inline pyomo/contrib/doe/fim_doe.py:996]: Remove any old code that was added just to get a specific example working.

File: pyomo/contrib/doe/fim_doe.py [hunk 11]
```diff
@@ -557,7 +575,7 @@ def optimize_doe(self,  design_values, if_optimize=True, objective_option='det',
             return analysis_square
 
     def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=None, specified_prior=None,
```
Comment (adowling2) [inline pyomo/contrib/doe/fim_doe.py:577]: This is a really big function. Does it make sense to divide it into a sequence of smaller private methods?

File: pyomo/contrib/doe/fim_doe.py [hunk 12]
```diff
@@ -623,6 +641,9 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
             # call generator function to get scenario dictionary
             scena_gen = Scenario_generator(self.param_init, formula=self.formula, step=self.step)
             scena_gen.generate_sequential_para()
+            scena_set = scena_gen.simultaneous_scenario()
+            self.scena_set = scena_set
+            print(self.scena_set)
 
             # if measurements are provided
             if read_output is not None:
```
Comment (adowling2) [inline pyomo/contrib/doe/fim_doe.py:649]: This code seems to be duplicated in the function. How can we refactor to remove the duplication?

File: pyomo/contrib/doe/fim_doe.py (hunk 11 above)
Comment (adowling2) [reply pyomo/contrib/doe/fim_doe.py:577]: How much of the workflow in this function is general across all supported modes? Or is most of this function series of if statements that have different sections for each mode? If the latter, we should strongly consider moving the code for each mode into a private method in the class. But let's do this AFTER identifying common code across the modes that can be generalized.

File: pyomo/contrib/doe/fim_doe.py [hunk 13]
```diff
@@ -970,11 +987,15 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
             zero_sens = np.zeros(len(self.param_name))
 
             # loop over measurement variables and their time points
-            for measurement_name in self.measure.model_measure_name:
+            for measurement_name in self.measure.model_measure_name_kaug:
```
Comment (adowling2) [inline pyomo/contrib/doe/fim_doe.py:990]: Is this inside a section of code only accessed in kaug mode?

File: pyomo/contrib/doe/fim_doe.py [hunk 14]
```diff
@@ -1053,12 +1129,16 @@ def __finite_calculation(self, output_record, scena_gen):
         for no_p, para in enumerate(self.param_name):
             # extract involved scenario No. for each parameter from scenario class
             involved_s = scena_gen.scenario_para[para]
+            #print('involveds:', involved_s)
```
Comment (adowling2) [inline pyomo/contrib/doe/fim_doe.py:1132]: Either remove these or add them as part of print_level 2 or 3.

File: pyomo/contrib/doe/reactor_kinetics_parmest.py [hunk 15]
```diff
@@ -0,0 +1,238 @@
+#################################################################################################################
```
Comment (adowling2) [inline pyomo/contrib/doe/reactor_kinetics_parmest.py:1]: The next step is to add tests similar to parmest.

File: pyomo/contrib/doe/fim_doe_tutorial_parmest.ipynb [hunk 16]
```diff
@@ -0,0 +1,791 @@
...
+    "\n",
+    "import numpy as np\n",
+    "#from scipy.interpolate import interp2d\n",
+    "import pandas as pd \n",
+    "from itertools import permutations, product, combinations\n",
+    "#import idaes\n",
+    "\n",
+    "from fim_doe import *\n",
+    "\n",
+    "import pyomo.common.unittest as unittest"
```
Comment (adowling2) [inline pyomo/contrib/doe/fim_doe_tutorial_parmest.ipynb:53]: This is not needed.

File: pyomo/contrib/doe/fim_doe_tutorial_parmest.ipynb [hunk 17]
```diff
@@ -0,0 +1,791 @@
+{
```
Comment (adowling2) [inline pyomo/contrib/doe/fim_doe_tutorial_parmest.ipynb:1]: This notebook should be divided into sections: (1) defining the model, (2) regressing the model parameters with Parmest, (3) performing MBDoE analysis with Pyomo.DOE. This notebook will become a tutorial for new Pyomo.DOE users.

**Review by adowling2 — COMMENTED** [review]
In the notebooks:
- Need to fix the error messages
- Need to remove the warning messages
- Need to fix Ipopt output that shows all iterations in restoration. Why is this happening?

**Comment by jialuw96** [comment]
> In the notebooks:
> 
> * Need to fix the error messages
> * Need to remove the warning messages
> * Need to fix Ipopt output that shows all iterations in restoration. Why is this happening?
> * Remove `import *` and instead name the functions being imported explicitly

In the fim_doe_tutorial_parmest.ipynb notebook the error message is resolved. Didn't commit the correct version before.

File: pyomo/contrib/doe/fim_doe.py (hunk 9 above)
Comment (jialuw96) [reply pyomo/contrib/doe/fim_doe.py:954]: ComponentUID

**Comment by adowling2** [comment]
@jialuw96 Can this PR get closed? Is it outdated? Closing a PR does not delete it. It just helps us remember this is no longer being actively developed.

**Pull request closed without merging**